### ISSUE TITLE
py_blake2*_new_impl produces inconsistent error messages, and raises OverflowError where ValueError might be better

### ISSUE THREAD

**Issue by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [python/cpython#74021]
BPO | [29835](https://bugs.python.org/issue29835)
--- | :---
Nosy | @tiran, @serhiy-storchaka, @zhangyangyu, @orenmn

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-03-17.12:08:00.055>
labels = ['type-bug', '3.7']
title = 'py_blake2*_new_impl produces inconsistent error messages, and raises OverflowError where ValueError might be better'
updated_at = <Date 2017-03-17.13:17:24.223>
user = 'https://github.com/orenmn'
```

bugs.python.org fields:
```python
activity = <Date 2017-03-17.13:17:24.223>
actor = 'christian.heimes'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2017-03-17.12:08:00.055>
creator = 'Oren Milman'
dependencies = []
files = []
hgrepos = []
issue_num = 29835
keywords = []
message_count = 2.0
messages = ['289757', '289759']
nosy_count = 4.0
nosy_names = ['christian.heimes', 'serhiy.storchaka', 'xiang.zhang', 'Oren Milman']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue29835'
versions = ['Python 3.7']
```

</p></details>

**Comment by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [comment]
```pycon
1. currently, py_blake2s_new_impl and py_blake2b_new_impl might produce
   inconsistent error messages (note that the first one happens on platforms
   where sizeof(long) > 4):
   >>> hashlib.blake2b(leaf_size=1 << 32)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   OverflowError: leaf_size is too large
   >>> hashlib.blake2b(leaf_size=1 << 1000)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   OverflowError: Python int too large to convert to C unsigned long
   
   >>> hashlib.blake2b(depth=256)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   ValueError: depth must be between 1 and 255
   >>> hashlib.blake2b(depth=256 << 1000)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   OverflowError: Python int too large to convert to C long
   
   there are similar inconsistent error messages when the function receives big
   and small out of range values for other arguments, too.
   
2. it might be better to raise a ValueError in the following cases:
   >>> hashlib.blake2b(leaf_size=-1)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   OverflowError: can't convert negative value to unsigned int

   >>> hashlib.blake2b(depth=-1 << 1000)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   OverflowError: Python int too large to convert to C long

   and maybe also in this case?
   >>> hashlib.blake2b(depth=1 << 1000)
   Traceback (most recent call last):
     File "<stdin>", line 1, in <module>
   OverflowError: Python int too large to convert to C long
```


   this might be considered as a sub-issue of bpo-29833.


note that solving the issue for leaf_size might be easier after bpo-29834 is
resolved, as one could add something like the following to the if block of
(leaf_size == (unsigned long) -1 && PyErr_Occurred()):

```py
               err = PyErr_Occurred();
               if (PyErr_GivenExceptionMatches(err, PyExc_OverflowError) ||
                   PyErr_GivenExceptionMatches(err, PyExc_ValueError)) {
                   PyErr_SetString(err,
                                   "leaf_size must be between 0 and 2**32-1");
               }
```


however, depth and other arguments are parsed by py_blake*_new, which is
generated by the argument clinic, so ISTM that solving the issue for them might
be harder.


(this issue was created while working on bpo-15988, as can be seen in the code
review comments of [PR 668](https://github.com/python/cpython/pull/668).)

**Comment by zhangyangyu** [comment]
I don't catch up with the progress but just FYI, the documentation of OverflowError states this situation: https://docs.python.org/3/library/exceptions.html#OverflowError. (I wanted to raise similar issues before ;-)

**Comment by picnixz** [comment]
FTR:

- Any parameter that fails to be parsed by AC leads to either an OverflowError or a ValueError (e.g., `leaf_size=-1` raises a ValueError but `leaf_size=1<<1000` raises an OverflowError).
- If the parsed parameters are outside a known range, they raise a ValueError, except (for legacy reasons) if `leaf_size > 0xFFFFFFFFU` or `node_offset > 0xFFFFFFFFFFFFULL` for BLAKE2s, in which case an OverflowError is raised.

Now, `leaf_size` is currently parsed as unsigned long, which is *at least* 4-byte long. On Windows, there won't be such exception but on Linux, the check will trigger. So, this is a way to make it consistent (namely, we raise OverflowError in both cases instead of asking for parsing an `uint32_t` which we could actually).

For `node_offset` it's a bit different as its maxvalue is $2^{48} - 1$. We could treat the valid range of it as $[0, 2^{48}-1]$, in which case we should raise a ValueError as for the other parameters.

Long story short:

- We have the clinic range: we raise a ValueError or an OverflowError (ValueError if conversion fails, OverflowError if it's outside the range).
- We have the parameter range: we raise a ValueError if it's outside that range. This makes the treatment of `node_offset` invalid as $2^{48}-1$ isn't a maxvalue for any C type.

We have a wider issue about the AC issue: https://github.com/python/cpython/issues/74020. cc @serhiy-storchaka

**Comment by serhiy-storchaka** [comment]
This is not just an AC issue. There are inconsistencies with using any C API -- see an old large issue #60192.

It would be better to raise the same exception (type and error message) consistently for all values out of the valid range: OverflowError if the range if implementation dependent, ValueError is it is implementation independent. But it is not always easy to do this using an existing C API. We need a new convenient C API, for use with `PyArgParse` and independently, which would convert to C integer and checked the range, raising the same exception (with customizable type and error message). It is difficult to design such API which would be flexible and simple to use. In any case, this is more general issue than just hashlib.blake2b().

**Comment by picnixz** [comment]
Does adding a new convenient C API for that requires a PEP? if so, I'm willing to write one. I had some idea with Argument Clinic where we would define something like:

```c
arg: numeric(c_type=int, min=10, max=123, error="custom error message")
```

If there is dependency between two args concerning their acceptable range, then the user is responsible for handling it in the non-autogenerated code with the appropriate exception. There shouldn't be error handling if we do not fit some exact C type.

For instance, we would first check if we can convert *arg* to the given *c_type*, raising OverflowError if this is not the case, then check if it's within some bounds (not necessarily tight ones) and raise ValueError if this is not the case, and for any other custom checks, the exception type should be a ValueError or in some very rare cases an OverflowError if this again depend on the underlying system.

**Comment by serhiy-storchaka** [comment]
Maybe. But if we know the range, and it is more narrow than the range of the C type, there should not be an OverflowError -- it should be ValueError. In most cases.

Argument Clinic is CPython-only. For users, there should be other ways -- `PyArg_Parse` related and standalone.

**Comment by picnixz** [comment]
> and it is more narrow than the range of the C type

This looks right. I'll try to see if I come up first with some specs.

> PyArg_Parse related and standalone.

My naive idea was to pass some structure defining the min/max etc, something like that:

```c
struct Info {
  size_t tsize;  // sizeof(OUT_TYPE)
  ssize_t smin;  // for signed min
  ssize_t smax;  // for signed max
  size_t min;    // for unsigned min
  size_t max;    // for unsigned max
  const char *err;
};

static struct Info myinfo = {...};
PyArg_Parse(args, "N", &info, &value);
``` 

We should either given smin or min and either smax or max. The tsize would help to know the expected output type size. We can extend this interface in the future if we add new fields to `Info` (maybe add some callback possible error callback function and success callback function, as well as some explicit bound checking function that takes either a size_t or a ssize_t depending on whether we expect a signed or unsigned value)
